### PR TITLE
Fix race condition

### PR DESCRIPTION
**Pull request by masahide**

### CODE REVIEW

**Comment by tagomoris** [comment]
"Race condition" for what?
And how can we reproduce it and check whether it's fixed or not?

**Comment by methane** [comment]
> "Race condition" for what?

* https://github.com/fluent/fluent-logger-golang/blob/e781f26187b040b56b1ce871bb4bbcca0d6abe5f/fluent/fluent.go#L166
* https://github.com/fluent/fluent-logger-golang/blob/e781f26187b040b56b1ce871bb4bbcca0d6abe5f/fluent/fluent.go#L183
* https://github.com/fluent/fluent-logger-golang/blob/e781f26187b040b56b1ce871bb4bbcca0d6abe5f/fluent/fluent.go#L198-L200
* https://github.com/fluent/fluent-logger-golang/blob/e781f26187b040b56b1ce871bb4bbcca0d6abe5f/fluent/fluent.go#L209-L211
* https://github.com/fluent/fluent-logger-golang/blob/e781f26187b040b56b1ce871bb4bbcca0d6abe5f/fluent/fluent.go#L228
* https://github.com/fluent/fluent-logger-golang/blob/e781f26187b040b56b1ce871bb4bbcca0d6abe5f/fluent/fluent.go#L249-L250

> And how can we reproduce it and check whether it's fixed or not?

There are no deterministic way to reproduce race condition.
But Go provides nice tool: race detector.

To use race detector, we should have stress test which runs racing point in parallel.
Personally, I feel fixing obvious, simple race with careful review is rational.
Because maintaining such test cases after fixing races has large cost.

File: fluent/fluent.go [hunk 1]
```diff
@@ -189,10 +196,9 @@ func (f *Fluent) Close() (err error) {
...
-	if f.conn != nil {
-		f.mu.Lock()
-		defer f.mu.Unlock()
-	} else {
+	f.mu.Lock()
+	defer f.mu.Unlock()
+	if f.conn == nil {
 		return
 	}
 	if f.conn != nil {
```
Comment (methane) [inline fluent/fluent.go]: This `if` statement is redundant.

File: fluent/fluent.go [hunk 2]
```diff
@@ -204,6 +210,8 @@ func (f *Fluent) close() (err error) {
 
 // connect establishes a new connection using the specified transport.
 func (f *Fluent) connect() (err error) {
+	f.mu.Lock()
+	defer f.mu.Unlock()
```
Comment (methane) [inline fluent/fluent.go]: `Dial` may take long time.  This lock shouldn't prevent write from other goroutines.

File: fluent/fluent.go [hunk 3]
```diff
@@ -220,44 +228,40 @@ func e(x, y float64) int {
...
-				waitTime := f.Config.RetryWait * e(defaultReconnectWaitIncreRate, float64(i-1))
-				time.Sleep(time.Duration(waitTime) * time.Millisecond)
+	for i := 0; ; i++ {
+		err := f.connect()
+		if err == nil {
+			f.mu.Lock()
+			f.reconnecting = false
+			f.mu.Unlock()
+			break
+		} else {
```
Comment (methane) [inline fluent/fluent.go]: Use `return` instead of `break`.
Then `else` can be removed.

**Comment by tagomoris** [comment]
#20 was fixed instead of this change. Thank you for contribution!

**Pull request closed without merging**